pkg/meta/sql.go

```diff
@@ -2637,35 +2647,170 @@ func (m *dbMeta) dumpEntry(inode Ino) (*DumpedEntry, error) {
 		return nil
 	})
 }
+func (m *dbMeta) dumpEntryFast(inode Ino) (*DumpedEntry, error) {
+	e := &DumpedEntry{}
+	n, ok := m.snap.node[inode]
+	if !ok {
+		return nil, fmt.Errorf("inode %d not found", inode)
+	}
+	attr := &Attr{}
+	m.parseAttr(n, attr)
+	e.Attr = dumpAttr(attr)
+	e.Attr.Inode = inode
 
-func (m *dbMeta) dumpDir(inode Ino, showProgress func(totalIncr, currentIncr int64)) (map[string]*DumpedEntry, error) {
-	var edges []edge
-	if err := m.engine.Find(&edges, &edge{Parent: inode}); err != nil {
-		return nil, err
+	rows, ok := m.snap.xattr[inode]
+	if ok && len(rows) > 0 {
+		xattrs := make([]*DumpedXattr, 0, len(rows))
+		for _, x := range rows {
+			xattrs = append(xattrs, &DumpedXattr{x.Name, string(x.Value)})
+		}
+		sort.Slice(xattrs, func(i, j int) bool { return xattrs[i].Name < xattrs[j].Name })
+		e.Xattrs = xattrs
 	}
+
+	if attr.Typ == TypeFile {
+		for indx := uint32(0); uint64(indx)*ChunkSize < attr.Length; indx++ {
+			c, ok := m.snap.chunk[fmt.Sprintf("%d-%d", inode, indx)]
+			if !ok {
+				return nil, fmt.Errorf("no found chunk target for inode %d indx %d", inode, indx)
+			}
+			ss := readSliceBuf(c.Slices)
+			slices := make([]*DumpedSlice, 0, len(ss))
+			for _, s := range ss {
+				slices = append(slices, &DumpedSlice{Chunkid: s.chunkid, Pos: s.pos, Size: s.size, Off: s.off, Len: s.len})
+			}
+			e.Chunks = append(e.Chunks, &DumpedChunk{indx, slices})
+		}
+	} else if attr.Typ == TypeSymlink {
+		l, ok := m.snap.symlink[inode]
+		if !ok {
+			return nil, fmt.Errorf("no link target for inode %d", inode)
+		}
+		e.Symlink = l.Target
+	}
+	return e, nil
+}
+
+func (m *dbMeta) dumpDir(inode Ino, tree *DumpedEntry, bw *bufio.Writer, depth int, showProgress func(totalIncr, currentIncr int64)) error {
+	bwWrite := func(s string) {
+		if _, err := bw.WriteString(s); err != nil {
+			panic(err)
+		}
+	}
+	var edges []*edge
+	var err error
+	var ok bool
+	if m.root == 1 {
+		edges, ok = m.snap.edges[inode]
+		if !ok {
+			return fmt.Errorf("no edge target for inode %d", inode)
+		}
+	} else {
+		if err := m.engine.Find(&edges, &edge{Parent: inode}); err != nil {
+			return err
+		}
+	}
+
 	if showProgress != nil {
 		showProgress(int64(len(edges)), 0)
 	}
-	entries := make(map[string]*DumpedEntry)
-	for _, e := range edges {
-		entry, err := m.dumpEntry(e.Inode)
+	if err := tree.writeJsonWithOutEntry(bw, depth); err != nil {
+		return err
+	}
+	sort.Slice(edges, func(i, j int) bool { return edges[i].Name < edges[j].Name })
+
+	for idx, e := range edges {
+		var entry *DumpedEntry
+		if m.root == 1 {
+			entry, err = m.dumpEntryFast(e.Inode)
+		} else {
+			entry, err = m.dumpEntry(e.Inode)
+		}
 		if err != nil {
-			return nil, err
+			return err
```

Comment:
Should we skip the edge and continue (log it as warning)?

Reply:
It has been modified